### PR TITLE
Update config parsing to handle new builder.py script format

### PR DESCRIPTION
**Pull request by RyanUnderhill**
Tested with Phi2 and the built in tests (includes updated genai_config.json files for tests)

Note: Config parsing will now give errors for unknown values in the config. So if a config loads without errors, all values were recognized.

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -16,120 +16,231 @@ ONNXTensorElementDataType TranslateTensorType(std::string_view value) {
...
+    } else if (name == "diversity_penalty") {
+      v_.diversity_penalty = static_cast<float>(value);
+    } else if (name == "length_penalty") {
+      v_.length_penalty = static_cast<float>(value);
+    } else
+      throw JSON::unknown_value_error{};
   }
 
   void OnBool(std::string_view name, bool value) override {
     if (name == "early_stopping") {
```
Comment (kunal-vaishnavi) [inline src/config.cpp:212]: Is this needed if `early_stopping = true` by default?

File: src/python/CMakeLists.txt [hunk 2]
```diff
@@ -10,7 +10,7 @@ set_target_properties(python PROPERTIES OUTPUT_NAME "onnxruntime_genai")
 
 if(USE_CUDA AND CMAKE_CUDA_COMPILER)
   set_target_properties(python PROPERTIES LINKER_LANGUAGE CUDA)
-  target_link_libraries(python PRIVATE cublasLt cublas cudnn curand cufft cudart)
+  # target_link_libraries(python PRIVATE cublasLt cublas cudnn curand cufft cudart)
```
Comment (yufenglee) [inline src/python/CMakeLists.txt:13]: why commenting out this line?

File: src/config.cpp [hunk 3]
```diff
@@ -16,120 +16,231 @@ ONNXTensorElementDataType TranslateTensorType(std::string_view value) {
...
  private:
-  Config::Model& model_;
+  Config::Model& v_;
+  EncoderDecoderInit_Element encoder_decoder_init_{v_.encoder_decoder_init};
+  Decoder_Element decoder_{v_.decoder};
 };
 
-struct Root_Element : JSON::Element {
-  explicit Root_Element(Config& config) : config_{config} {}
+struct Search_Element : JSON::Element {
```
Comment (yufenglee) [inline src/config.cpp:175]: nit: name it as Generato_Element for consistency

File: test/test_models/hf-internal-testing/tiny-random-gpt2-fp16/genai_config.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+  "model": {
+    "type": "gpt2",
+    "pad_token_id": 98,
+    "bos_token_id": 98,
+    "eos_token_id": 98,
+    "vocab_size": 1000,
+    "logits_type": "float16",
```
Comment (yufenglee) [inline test/test_models/hf-internal-testing/tiny-random-gpt2-fp16/genai_config.json:8]: Could we add the capability to infer the type from model?

File: src/python/CMakeLists.txt (hunk 2 above)
Comment (RyanUnderhill) [reply src/python/CMakeLists.txt:13]: Oops, I didn't mean to check this in.

File: src/config.cpp (hunk 1 above)
Comment (RyanUnderhill) [reply src/config.cpp:212]: If you want to set it to false, perhaps?

File: test/test_models/hf-internal-testing/tiny-random-gpt2-fp16/genai_config.json (hunk 4 above)
Comment (RyanUnderhill) [reply test/test_models/hf-internal-testing/tiny-random-gpt2-fp16/genai_config.json:8]: Yep, that'll be in the next PR. This unblocks people using the updated builder.py

File: src/config.cpp (hunk 3 above)
Comment (RyanUnderhill) [reply src/config.cpp:175]: Hmm, the values here are for how to do the search (beamsearch/greedysearch).
But I can see how they're also how the generation loop works.
But the GeneratorParams also has the model inputs.
Up to you, I can call this Generator_Element, but it's not directly related to GeneratorParams.

**Review by yufenglee — APPROVED** [review]

**Pull request merged**